Refactor index-tuple implementation

mphe opened the pull request:
Some code operates on two arbitrary points, e.g. for constraint handling in PointArray. These point tuples have been represented as Array[int] so far.
This PR replaces these usages of Array[int] with Vector2i for several reasons:
  - Less error prone
  - No more problems with typed arrays
  - Better suits the tuple type: there are exactly two int elements, not more, not less, nothing else.

`tuple.gd` contains utility functions for dealing with said index tuples as they have the special equality behavior `T(x, y) <=> T(y, x)`, which needs to be respected when searching in arrays or dictionaries using this tuple type.
Unfortunately, Godot does not allow custom equality and hash functions, so we have to write our own accessor functions for arrays and dictionaries.
This patch improves the previous implementations by
- Refactoring existing functions
- Providing a complete set of accessor functions for arrays and dictionaries
- O(1) access for dictionaries rather than O(n)
- Code cleanup / More straightforward way of working with index tuples.

Finally, as I already mentioned in Discord, I fixed a bug in IndexMap.indicies_to_edges(): The function now correctly operates on the array's values instead of only its indices.

---

This PR introduces quite a bit of refactoring and needs more testing.
Should fix #150 (again) as the problematic parts no longer exist.

@araslanix if you find some time, could you test this PR on your project? You seem to have a project that hits many edge cases, which is insanely valuable here.

I'm also planning to replace various occurences of Array[int] (in non tuple use cases) with PackedInt32Array for performance reasons and also to get rid of as much typed arrays as possible without sacrificing static typing.
Might add this as another commit here or in a new PR, depending on how long it takes.

araslanix commented:
Update: Opening scenes that use SS2D in the Godot editor once, saving, and then running the project appears to resolve the issue. The git diff shows that some conversions and new lines have been added to the point array resource. 
![image](https://github.com/SirRamEsq/SmartShape2D/assets/47268404/58c2d616-b09b-4de5-afb2-7dee93d6c55a)
I keep them in a separate .tres file since their references change every time I open a scene which makes it difficult to see and commit my actual changes. will test more and update if I see anything.

mphe commented:
It's fine, keep posting here.

I thought I had fixed the fill texture bug...
Could you provide a minimal reproduction scene/project? Basically just a scene with the buggy shape. Possibly even the point array resource should suffice.

araslanix commented:
Sure, here's a link to a shape that can't load the placeholder texture:
https://drive.google.com/file/d/1CqsdEz2vlcQgqsZx61OO_2YOOAuW92Dl/view?usp=drive_link

As mentioned, I don't have access to my tools, but without updating the plugin, it should look like this:
![image](https://github.com/SirRamEsq/SmartShape2D/assets/47268404/7a820222-ca84-408d-ac49-25df3a2b362a)

limbonaut commented:
One of my test scenes got broken with this PR - most likely same bug as @araslanix mentioned.
[test_mirrored.zip](https://github.com/user-attachments/files/15595012/test_mirrored.zip)

It seems that the first and last keys got disconnected on load. They occupy the same spot, but are no longer considered joined. So the shape thinks it is not closed.

mphe commented:
Yes, I had the same issue in my tests already and thought I fixed it. Guess there is another edge case I have overlooked.
I'll look into it.

limbonaut commented on addons/rmsmartshape/shapes/point_array.gd:
```diff
@@ -526,71 +512,66 @@ func debug_print() -> void:
-			push_error("Material Override Dictionary KEY is not an Array with 2 points!")
-		if not dict[k] is SS2D_Material_Edge_Metadata:
-			push_error("Material Override Dictionary VALUE is not SS2D_Material_Edge_Metadata!")
+	SS2D_IndexTuple.dict_validate(dict, SS2D_Material_Edge_Metadata)
```
Aha, now I see why.

limbonaut commented:
LGTM, besides those small nitpicks. I find it hard sometimes to follow inferred types, like `var x := func()`, especially while reading code on GH when you can't inspect a symbol.

mphe commented:
> Yes, I had the same issue in my tests already and thought I fixed it. Guess there is another edge case I have overlooked. I'll look into it.

I'm dumb and forgot to commit those changes...
It should be fixed now.

mphe commented on addons/rmsmartshape/lib/tuple.gd:
```diff
@@ -1,48 +1,139 @@
+	if value != null:
+		return value
+
+	return dict.get(flip_elements(tuple))
```
Should be resolved now with yesterday's amend. This was part of why shapes sometimes weren't recognized as closed.

mphe commented on addons/rmsmartshape/shapes/index_map.gd:
```diff
@@ -4,90 +4,85 @@ class_name SS2D_IndexMap
-			var highest: Array[int] = final_segments[idx_highest]
+			var idx_lowest := to_join_tuple[0]
+			var idx_highest := to_join_tuple[1]
+			var lowest := final_segments[idx_lowest]
```
I can see how it is more readable with explicit types, but I find implicitly inferred types more reliable.
But I can add explicit types here.

mphe commented:
Fixed the requested changes. I'll squash the commits when the PR is approved and ready.
@araslanix Please let me know if it now works for you and if you find any further problems :)

araslanix commented:
> Fixed the requested changes. I'll squash the commits when the PR is approved and ready.
> @araslanix Please let me know if it now works for you and if you find any further problems :)
> 

Ah, that was quick! Nice work. I'll be back on Saturday 15th, will do a full check on everything when I'm back. But if you need to merge already, you can go ahead with the shape I sent. If it works then probably it's all good 👍

limbonaut commented:
One more issue I found, but it's also present in the `master`: adding a new `SS2D_Shape` with create tool selected spams the following error non-stop:
```
  scene/resources/curve.cpp:700 - Index p_index = 0 is out of bounds (points.size() = 0).
  res://addons/rmsmartshape/shapes/point_array.gd:658 - Out of bounds set index '0' (on base: 'PackedVector2Array')
  scene/resources/curve.cpp:700 - Index p_index = 0 is out of bounds (points.size() = 0).
  res://addons/rmsmartshape/shapes/point_array.gd:658 - Out of bounds set index '0' (on base: 'PackedVector2Array')
  scene/resources/curve.cpp:700 - Index p_index = 0 is out of bounds (points.size() = 0).
  res://addons/rmsmartshape/shapes/point_array.gd:658 - Out of bounds set index '0' (on base: 'PackedVector2Array')
```

mphe commented:
> One more issue I found, but it's also present in the `master`: adding a new `SS2D_Shape` with create tool selected spams the following error non-stop:

Fixed it.

I also changed various occurences of Array[int] with PackedInt32Array.
There were a few instances of PackedInt64Array which I changed to Int32 because it should suffice I think.

limbonaut commented:
In my tests in an existing project, unless I resave a scene with the updated addon, it loads in-game with fill texture empty. So each scene needs to be resaved with the new version. For some reason, I also get `convex decomposition failed` errors. If I resave the scene, those errors go away, and fill texture rendered as expected. Should we consider this a bug?

mphe commented:
I also experienced this problem and I'm unsure why it happens. I also immediately forgot about it because it is fixed upon saving. I can try to investigate some more.
With the goal of maintaining backwards compatibility, it should probably be considered as a bug.

mphe commented:
Actually, I can't reproduce it anymore. Have you tested the most recent version of this branch? Could you share a test scene?

I stumbled over another problem in your `test_mirrored.tscn` scene, where `PackedInt64Array`s are used as constraint keys. I'm not sure how that happened, because these should have been `Array`s all along, but nevertheless I added backwards support for Packed Arrays there, too.

limbonaut commented:
You just need to create a shape using the master branch, and run it using this PR's branch. However, the trick is not to open that scene in the editor - or it will be resaved on "Run". I just assigned it in project settings in `Application -> Run -> Main Scene`, and ran the ss2d project.

Here's a scene that breaks for your convenience:
[test_case_fill.zip](https://github.com/user-attachments/files/15756947/test_case_fill.zip)

limbonaut commented:
> I stumbled over another problem in your `test_mirrored.tscn` scene, where `PackedInt64Array`s are used as constraint keys. I'm not sure how that happened, because these should have been `Array`s all along, but nevertheless I added backwards support for Packed Arrays there, too.

It looks like `test_mirrored` is broken in master branch too. Maybe, it got corrupted somehow.

limbonaut commented:
I think it's this bit:
```
func _set_constraints(cs: Dictionary) -> void:
	_constraints = cs

	# For backwards compatibility (Array to Vector2i transition)
	if Engine.is_editor_hint():
		SS2D_IndexTuple.dict_validate(_constraints, TYPE_INT)

```
Conversion is performed in editor. Whether it's a bug or feature, I don't know. It kinda would eat up cycles at runtime. But if left like that, there will be questions about shapes breaking in existing projects.

limbonaut commented:
What if we add a version integer to point_array, check it on load, and perform conversion if needed?

mphe commented:
Ah good catch. I think it should run during runtime, not only in editor.
The point array version number sounds like a good idea but a little overkill for this scenario. 
I would keep it in there as it is a negligible one-time cost (I can also run some time measures when I'm back at home). 
At some point we can make a release that breaks backwards compatibility and remove it there.

limbonaut commented:
> At some point we can make a release that breaks backwards compatibility and remove it there.

I'd add a comment with a reminder to remove it later, maybe dated for removal next year or something.

mphe commented:
Fixed it.

> UPDATE: And I don't get `convex decomposition failed` with that scene for some reason,

I think that might have been related to the constraint issue that I forgot to commit at the beginning.

limbonaut commented:
So I measured combined scene load time increase for every shape checking constraints in a relatively big level, and it's +0.2ms. Works for me.

limbonaut commented:
Thanks, LGTM

mphe commented:
> So I measured combined scene load time increase for every shape checking constraints in a relatively big level, and it's +0.2ms. Works for me.

That's fine. More than I expected but negligible for a one time cost.
There has also been dict validation before this PR to check for type consistency, so it should not really change anything.

mphe commented:
Thanks for testing. Is the randomizing issue related to this PR or does it also occur with earlier versions, e.g. the current release on the Asset Library?

araslanix commented:
No worries. With these settings and the asset library version, I can fit same-sized textures into tighter spaces. However, in the current master version, slightly more space is needed between points. Unclear whether this is due to this PR or some previous one, as the textures are random and change with each reload. However, I observed this while comparing the asset library version to the current master.
![image](https://github.com/SirRamEsq/SmartShape2D/assets/47268404/2ae6f556-6a01-4c13-8a9f-8ac5ddea434f)

mphe commented:
Created an issue #157 to track the texture randomization bug.
